hs/src/Reach/Eval/ImportSource.hs

```diff
@@ -0,0 +1,511 @@
+
+toBase16 :: ByteArrayAccess b => b -> T.Text
+toBase16 a = decodeUtf8
+  . convertToBase Base16
+  $ (convert a :: B.ByteString)
+
+
+fromBase16 :: ByteArray b => T.Text -> Either T.Text b
+fromBase16 = either (Left . T.pack) Right
+  . convertFromBase Base16
+  . encodeUtf8
+
+
+newtype SHA = SHA (Digest SHA256)
+  deriving (Eq, Show, Ord)
+
+instance ToJSON SHA where
+  toJSON (SHA d) = toJSON $ toBase16 d
+
+instance FromJSON SHA where
+  parseJSON = withText "SHA" $ \a -> maybe
+    (fail $ "Invalid SHA: " <> show a)
+    (pure . SHA)
+    (digestFromByteString =<< either (const Nothing) Just (fromBase16 @B.ByteString a))
+
+instance ToJSONKey   SHA
+instance FromJSONKey SHA
+
+
+data GitSaas = GitSaas
+  { acct ::  String
+  , repo ::  String
+  , ref  ::  String
+  , dir  :: [FilePath]
+  , file ::  FilePath
+  } deriving (Eq, Show, Generic, ToJSON, FromJSON)
+
+
+data HostGit
+  = GitHub    GitSaas
+  | BitBucket GitSaas
+  deriving (Eq, Show, Generic, ToJSON, FromJSON)
+
+
+data ImportSource
+  = ImportLocal     FilePath
+  | ImportRemoteGit HostGit
+  deriving (Eq, Show)
+
+
+-- | Represents an entry indexed by 'SHA' in 'LockFile' capturing details by
+-- which a directly-imported package module may be fetched.
+--
+-- This type tracks parsed but unevaluated @git@ imports in its 'host' field.
+--
+-- __The /unevaluated/ distinction here is important:__ e.g. if the user asks
+-- to import a module from the package's @master@ branch, in which case we need
+-- to perform some additional steps to fix the import to a specific @git SHA@.
+-- This is why the 'refsha' field exists.
+--
+-- Transitive dependencies which are local to the direct import in 'host' and
+-- which belong to the same 'refsha' are discovered and added the 'ldeps'
+-- field.
+data LockModule = LockModule
+  { host   :: HostGit            -- ^ Raw result of parsing import statement
+  , refsha :: HostGitRef         -- ^ Git SHA to which @host@ refers at time of locking
+  , uri    :: GitUri             -- ^ A @git clone@-able URI
+  , ldeps  :: M.Map FilePath SHA -- ^ Repo-local deps discovered during @gatherDeps_*@ phase
+  } deriving (Eq, Show, Generic, ToJSON, FromJSON)
+
+
+data LockFile = LockFile
+  { version :: Int
+  , modules :: M.Map SHA LockModule
+  } deriving (Eq, Show, Generic, ToJSON, FromJSON)
+
+
+lockFileEmpty :: LockFile
+lockFileEmpty =  LockFile
+  { version = 1
+  , modules = mempty
+  }
+
+
+--------------------------------------------------------------------------------
+
+data PkgError
+  = PkgGitCloneFailed         String
+  | PkgGitCheckoutFailed      String
+  | PkgGitFetchFailed         String
+  | PkgGitRevParseFailed      String
+  | PkgLockModuleDoesNotExist FilePath
+  | PkgLockModuleShaMismatch  FilePath
+  | PkgLockModuleUnknown      HostGit
+  | PkgLockModifyUnauthorized
+  deriving (Eq, ErrorMessageForJson, ErrorSuggestions)
+
+
+instance Show PkgError where
+  show (PkgGitCloneFailed         s) = "`git clone` failed: "          <> s
```

Comment:
Style nit; I really like `show = \case` because you can have a shared `while` at the bottom and you don't repeat the name of the function